Controllers films users

i2GR opened the pull request:
SP9 FP initial by Igor I.

i2GR committed:
new empty filmorate:
added maven, junit, add all to git

i2GR committed:
added logback.xml with configuration
(previously log4j2 tried)

i2GR committed:
added validations
added backlog config

i2GR committed:
finilized classes
unit tests added

i2GR committed:
Modified validators and controllers to pass Postman test
log warning moved to validators
tests checked

i2GR committed:
SP9 FP initial
by Igor I.

afomkina commented on pom.xml:
```diff
@@ -0,0 +1,81 @@
+	<artifactId>filmorate</artifactId>
+	<version>0.0.1-SNAPSHOT</version>
+	<name>filmorate</name>
+	<description>YP SP9 final project (ii)</description>
```
Лучше добавить в описание красткий тезис чем занимается приложение

i2GR replied:
добавлено описание в тег description

afomkina commented on .gitignore:
```diff
@@ -43,4 +43,15 @@ build/
+**/target/
+
+!**/src/test/
+**/target/
```
Нужно добавить игнорирование для `.mvn/wrapper`, `mvnw` и `mvnw.cmd`

i2GR replied:
спасибо.
дополнительно прочитано, почему так...

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,61 @@
+    private boolean customValidation = false;
+    private final Map<Integer, Film> idMapFilm = new HashMap<>();
+
+    @PostMapping("")
```
Если не требуется дополнить ссылку запроса, то лучше убрать параметры у аннотации:
```suggestion
    @PostMapping
```
Поправь, пожалуйста, во всех методах контроллеров

i2GR replied:
недосмотрел...

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,61 @@
+@RequestMapping("/films")
+public class FilmController {
+    private int id = 0;
+    @Setter
```
Здесь сеттер не нужен, лучше сделать переменную `customValidation` финальной.

i2GR replied:
boolean customValidation используется только для тестов ручной валидации, поэтому в контексте других замечаний это поле удалено из классов

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,67 @@
+
+    private int id = 0;
+
+    @Setter
```
Нужно поправить как в классе FilmController

i2GR replied:
исправлен code-style (пустые строки)

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,61 @@
+    public Film addNewMovie(@Valid @RequestBody Film film) throws ValidationException {
+        log.info("requested movie add");
+        // здесь валидатор остался только для написанных тестов
+        if (customValidation) film = FilmValidator.validate(film);
```
Оставлять код, чтобы тесты проходили корректно - небезопасно. Нужно избавиться от дублирующий валидации

i2GR replied:
+

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,61 @@
+public class FilmController {
+    private int id = 0;
+    @Setter
+    private boolean customValidation = false;
```
Лучше переименовать в 
```suggestion
    private boolean isValid = false;
```
Так код будет проще читать.

i2GR replied:
boolean customValidation используется только для тестов ручной валидации, поэтому в контексте других замечаний это поле удалено из классов

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,61 @@
+
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.web.bind.annotation.*;
```
_Саморазвитие_
В коммерческой разработке не принято использовать `wildcard` в импортах.
Вот отличная [статья](https://struchkov.dev/blog/using-wildcard-imports-in-java/) почему и как настроить идею, чтобы избавиться от `*`
Кстати, в IDEA есть отличное сочетание клавиш `Ctrl + Alt + O` - при нажатии сочетания в нужном классе автоматически оптимизируются импорты и удаляются ненужные

i2GR replied:
раз принято - значит принято.
я сам не люблю неопределенности:)
хотя в приведенной статье говорится об обратном со ссылкой на "чистый код":
J1: Используйте обобщенные директивы импорта "даже в унаследованном коде лучше
применять обобщенный импорт". Мартин Р. М29 Чистый код: создание, анализ и рефакторинг. - Спб.: Питер, 2019. стр 347

afomkina replied:
Да, но ниже расписана мысль автора статьи, что большое количество импортов сейчас не является проблемой, так как среда разработки скрывает это полотно от разработчиков, пока ему не потребуется туда заглянуть. А вот если потребуется заглянуть, гораздо удобнее и проще понимать, какие импорты были добавлены и используются данным классом, когда они все явно прописаны, а не схлопнуты в wildcard

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,28 @@
+    private String email;
+
+    @NotNull(message = "User.login is null")
+    @Pattern(regexp = "^\\w{1,}$", message = "User.login contains non letter o digit symbols")
```
🔥

i2GR replied:
даже добавлять в коммент :fire: приходится учиться
опечатка "o digit" одна из причин переписать тесты

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,28 @@
+
+    private String name;
+
+    @BirthDayPast
```
Можно заменить на уже реализованную аннотацию [@PastOrPresent ](https://www.baeldung.com/javax-validation#:~:text=%40Past%20and-,%40PastOrPresent,-validate%20that%20a) - она валидирует, что дата должна быть в прошлом или не позже, чем сейчас

i2GR replied:
какая то магия :eyes:, если оно само валидирует поле типа LocalDate c незаданным форматом

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,49 @@
+    FilmController fc = new FilmController();
+
+    {
+        fc.setCustomValidation(true);
```
Лучше не создавать сеттеры или какой-то другой код специально для тестов. Такой подход делает программу менее безопасной и более уязвимой к нечаянным ошибкам

i2GR replied:
действительно... спасибо.
я об этом недостаточно подумал, специально введ поле customValidation, но только потому что:
а) раз уж сначала был создан ручной валидатор, нужно было предусмотреть "демонстрацию его работы", по крайней мере в тестах
б) создание поля-статуса было наилегчайшим путем включить его в класс контроллера и убрать его ,т.к.
в) использование валидации через аннотации делает ненужным ручную валидацию и все что с ней связано включая тесты нужно будет после замечаний убирать. А замечания были ожидаемы...

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Test
+    void addNewMovie() throws ValidationException {
+        assertEquals(film,fc.addNewMovie(film));
```
Лучше продумать тест-кейсы таким образом, чтобы нам не потребовался какой-либо доступ к полям тестируемого класса.
Например, при добавлении фильма, метод `addNewMovie` возвращает нам объект `film`, если добавление прошло успешно или бросает ошибку, если нет.
В первом тесте можно передать методу фильм с корректно заполненными полями и проверить, что полученный фильм равен фильму, который мы передавали.
А во втором тесте, можно передать фильм с пустым полем "название" и проверить, что мы получили ошибку при выполнении данного метода. И так далее.
Перепиши, пожалуйста тесты таким образом, чтобы никак не взаимодействовать с полями тестируемого класса

i2GR replied:
если я правильно понял ,то в тесты добавлены билдеры фильмов/юзеров
в каждом тесте используется новый экз. фильма/юзера из билдера
ВОПРОС:
не очень понятно, как в рамках одного теста без доступа к полям простым способом создать копию фильма (или юзера) с измененными полями - модифицированный объект, чтобы его обновить в контроллере.
прямой путь - новый экземпляр с частично повторяющимися полями

afomkina replied:
Тут, наверное, я больше хотела обратить внимание на валидацию, которая была отставлена для тестов и предложить варианты решения, как можно обойтись без нее. 
Пример выше предлагает создать экземпляр класса `FilmController` и проверять сами методы, на разных данных по фильмам и юзерам. 
Видимо, слишком сильно мне запало в душу это поле customValidation 😁

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,48 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class UserControllerTest {
```
Поправь, пожалуйста, тесты для пользователей в соответствии с комментариями для тестов фильмов

i2GR replied:
исправлено... (надеюсь)

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,49 @@
+class FilmControllerTest {
+
+    private final Film film = ItemForTest.setDefaultTestFilm(1);
+    FilmController fc = new FilmController();
```
Не хватает модификатора доступа

i2GR replied:
тесты переписаны (заново)

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/model/FilmTest.java:
```diff
@@ -0,0 +1,145 @@
+        film.setDescription(null);
+
+        Set<ConstraintViolation<Film>> constraintViolations = validator.validate( film );
+        List<String> violationMessages = constraintViolations.stream().map(v -> v.getMessage())
```
Стримы лучше бить на несколько строк, так их удобнее читать:
```suggestion
        List<String> violationMessages =
            constraintViolations.stream()
                .map(v -> v.getMessage())
                .collect(Collectors.toList());
```

i2GR replied:
исправлено

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/model/FilmTest.java:
```diff
@@ -0,0 +1,145 @@
+        film.setDescription("");
+
+        Set<ConstraintViolation<Film>> constraintViolations = validator.validate( film );
+        List<String> violationMessages = constraintViolations.stream().map(v -> v.getMessage())
```
Можно вместо лямбды  `v -> v.getMessage()` 
использовать [method reference](https://www.baeldung.com/java-method-references): `ConstraintViolation::getMessage` 
везде, где это возможно

i2GR replied:
исправлено

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/model/FilmTest.java:
```diff
@@ -0,0 +1,145 @@
+
+class FilmTest {
+    private static Validator validator;
+    private static Film film;
```
Лучше не делать поле статичным, если того не требует контекст. Вот отличная [статья](https://medium.com/att-israel/should-you-avoid-using-static-ae4b58ca1de5), почему стоит избегать `static` везде, где это возможно

i2GR replied:
исправлено

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/model/UserTest.java:
```diff
@@ -0,0 +1,103 @@
+class UserTest {
+
+    private static Validator validator;
+    private static User user;
```
Лучше убрать у поля `static`

i2GR replied:
исправлено

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/model/FilmTest.java:
```diff
@@ -0,0 +1,145 @@
+    private static Validator validator;
+    private static Film film;
+
+    @BeforeAll
```
Лучше отказаться от `@BeforeAll` и перенести инициализацию валидации в блок `@BeforeEach`.

i2GR replied:
исправлено

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/model/UserTest.java:
```diff
@@ -0,0 +1,103 @@
+    private static Validator validator;
+    private static User user;
+
+    @BeforeAll
```
Лучше перенести в `@BeforeEach`

i2GR replied:
исправлено

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/validation/ValidationException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.yandex.practicum.filmorate.controller.validation;
+
+public class ValidationException extends Exception{
```
Исключение должно быть непроверяемым, следует наследовать этот класс от `RuntimeException`. Так же следует исключить его из заголовков всех методов, так как непроверяемые исключения не нужно в них указывать)

i2GR replied:
исправлено, но
ВОПРОС: почему лучше исключение должно быть непроверяемым?

afomkina replied:
Потому что подобный тип ошибок обрабатывается JVM во время работы программы и никак не влияет на компиляцию. Второй момент - их не нужно явно отлавливать и обрабатывать - меньше шаблонного кода и проблем, которые могу возникнуть по невнимательности. Ну и самое мое любимое - так принято в коммерческой разработке 😁
Вообще тема очень холиварная и есть два прямо противоположных мнения, как правильно делать.
Но так как у нас в проекте не заложена какая-либо сложная логика обработки ошибок валидации, я считаю, что лучше воспользоваться методом наименьшего сопротивления и не плодить множественные try-catch/ throws, когда это не нужно

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/validation/ValidationExceptionHandler.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    @ExceptionHandler(ValidationException.class)
+        public ResponseEntity<Object> handleValidationException(ValidationException ve, WebRequest request) throws Exception {
+        log.warn(ve.getMessage());
```
Лучше добавить в логирование какое-либо сообщение о том, что произошло, а с помощью заполнителя (placeholder)`{}`, добавить в сообщение лог ошибки.
Например:
```suggestion
        log.warn("Некорректно заполнены поля body: {}", ve.getMessage());
```

i2GR replied:
исправлено

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/validation/ValidationExceptionHandler.java:
```diff
@@ -0,0 +1,54 @@
+        try {
+            return handleValidationException(new ValidationException(ex.getMessage()), request);
+        } catch (Exception e) {
+            //throw new RuntimeException(e); // предложено идеей
```
Лучше удалять ненужный код, а не комментировать его, так как код становится менее читаемым и отвлекает разработчиков. Если он вдруг потребуется его всегда можно восстановить из гита

i2GR replied:
исправлено

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/validation/ValidationExceptionHandler.java:
```diff
@@ -0,0 +1,54 @@
+ */
+@Slf4j
+@ControllerAdvice
+public class ValidationExceptionHandler extends ResponseEntityExceptionHandler {
```
Лучше вынести кастомные ошибки и хэндлер в отдельный пакет

i2GR replied:
в пакете controller.validation после удаления "ручных" валидаторов остались только кастомное исключение и валидоры через аннотации и сами аннотация (сама аннтоация -одна осталась)

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/validation/StringValidator.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * общий класс с методами фалидации строк согласно первой части ТЗ
+ */
+public class StringValidator {
```
Лучше удалить этот класс, так как поля проверяются с помощью аннотаций валидации, не стоит хранить дублирующие методы

i2GR replied:
+
он не нужен как и все "ручные" валидаторы

afomkina commented on pom.xml:
```diff
@@ -0,0 +1,81 @@
+
+		<dependency>
+			<groupId>org.junit.platform</groupId>
+			<artifactId>junit-platform-suite-commons</artifactId>
```
Лишняя зависимость

i2GR replied:
этого я не знал...

afomkina commented on pom.xml:
```diff
@@ -0,0 +1,81 @@
+
+		<dependency>
+			<groupId>org.hibernate.validator</groupId>
+			<artifactId>hibernate-validator</artifactId>
```
Лишняя зависимость

i2GR replied:
и этого я не знал...

afomkina reviewed (changes requested):
Хорошая работа - молодец 👍
Из общих комментариев: нужно поправить весь проект по `code-style`: это легко можно сделать, если использовать горячие клавиши в IDEA [Ctrl+Alt+L](https://www.jetbrains.com/idea/guide/tips/reformat-file/)
Важно не использовать сокращения и однобуквенные переменные в проекте, так как в коммерческой разработке такой подход считается code-smell и делает код менее читаемым

i2GR committed:
SP9 FP review 1 fix
fix based review comments
controllers refactoring
logback.xml modified (one controller log)
by Igor I.

i2GR commented:
Спасибо за ревью!
замечания исправлены но есть пара вопросов (в комментах)
- вопрос про обобщенные директивы импорта (в чистом коде, наоборот, говорится, что можно)
- не очень понятно, как в рамках одного теста без доступа к полям простым способом создать копию фильма (или юзера) с измененными полями - модифицированный объект, чтобы его обновить в контроллере.
прямой путь - новый экземпляр с частично повторяющимися полями
- почему лучше исключение должно быть непроверяемым?

afomkina commented:
> Спасибо за ревью! замечания исправлены но есть пара вопросов (в комментах)
> 
> * вопрос про обобщенные директивы импорта (в чистом коде, наоборот, говорится, что можно)
> * не очень понятно, как в рамках одного теста без доступа к полям простым способом создать копию фильма (или юзера) с измененными полями - модифицированный объект, чтобы его обновить в контроллере.
>   прямой путь - новый экземпляр с частично повторяющимися полями
> * почему лучше исключение должно быть непроверяемым?

Ответила более подробно в тредах под самими вопросами, здесь продублирую тезисно:

1) Самая большая проблема wildcard в том, что можно легко пропустить конфликт имен, а потом долго и упорно искать почему же не запускается программа - это очень неприятно и часто отнимает много времени
2) Тут скорее просто не там тегнула комментарий и немного тебя сбила с мысли, этот комментарий больше был про поле оставленное для тестирование
3) В нашем случае нет смысла в проверяемом исключении, так как у нас нет сложной логики обработки ошибок валидации

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/validation/FixedValues.java:
```diff
@@ -0,0 +1,11 @@
+ * класс, содержащий конcтанты
+ */
+public abstract class FixedValues {
+    public static final int MOVIE_DESCR_LENGTH = 200;
```
Поле больше нигде не используется, поэтому его лучше удалить

i2GR replied:
а может быть лучше исправить таким образом чтобы избавиться от "магических чисел":
- и использовать константу в аннотации
- использовать константу в тестах длины поля Description (и переименовать методы)

`class Film 
//
    @NotBlank(message = "Film.description is blank")
    @Size(max = FixedValues.MAX_MOVIE_DESCR_LENGTH, message = "Film.description too long")
    private String description;
//`

afomkina replied:
Хорошая идея 👍

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/BasicController.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    protected T addNew( T item) {
+        log.info(String.format("requested %s to add", dtoName));
```
При логировании параметров, лучше использовать заполнитель(placeholder) `{}`. На его место автоматически подставится нужное значение и не придется использовать дополнительный метод форматирования:
```suggestion
        log.info("requested {} to add", dtoName);
```

i2GR replied:
недосмотрел... стандартная же фича логгера

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/BasicController.java:
```diff
@@ -0,0 +1,54 @@
+            log.info(dtoName + " added");
+            return item;
+        }
+        throw new ValidationException(dtoName + " to add already exists");
```
А вот тут удобнее будет использовать форматирование, а не конкатенацию

i2GR replied:
исправлено

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,27 @@
+    private String name;
+
+    @PastOrPresent(message = "User.birthdate is in Future")
+    LocalDate birthday;
```
Не хватает модификатора доступа

i2GR replied:
эх, ошибка уровня первый спринт:)

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/controller/ControllerTest.java:
```diff
@@ -0,0 +1,55 @@
+        this.controller = controller;
+    }
+
+    @Test()
```
Лишние скобочки

i2GR replied:
как так то....

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/controller/ControllerTest.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Test()
+    void addNewNull() {
+        assertThrows(Exception.class, () -> controller.addNew(null));
```
Лучше проверять, что получили конкретное исключение, которое и ждем, как в тесте ниже, а не обобщенный `Exception`

i2GR replied:
недосмотрел...

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/controller/ControllerTest.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    void getAll(T original) {
+        assertEquals(0, controller.getAll().size());
```
Можно упростить проверку до:
```suggestion
          assertTrue(controller.getAll().isEmpty());
```
Метод `isEmpty()` возвращает `true`, если `размер списка == 0` или `false` в любом другом случае

i2GR replied:
да так проще. действительно.
просто я встречал ситуации ,когда IDEA часто предлагала assertEquals() вместо assertTrue()
сейчас я наверное понял почему. кажется она настолько умна, чтобы предусматривать, что будет легче читаться.
assertTrue(controller.getAll().isEmpty()) здесь получается легче человеко-читаемо. буду проверять дальше эту теорию

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/controller/FilmControllerTest.java:
```diff
@@ -0,0 +1,46 @@
+
+    private Film film;
+
+    public FilmControllerTest() {
```
Лучше сделать `private`, так как больше нигде не понадобится.

i2GR replied:
это фиаско

afomkina commented on src/test/java/ru/yandex/practicum/filmorate/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,46 @@
+
+    private User user;
+
+    public UserControllerTest() {
```
Лучше сделать `private`

i2GR replied:
надо быть внимательнее

afomkina reviewed (changes requested):
Молодец, что оставляешь вопросы и стараешься разбираться 👍

i2GR commented:
Спасибо за ревью №2
прямо стыдно :) за часть очень грубых ошибок по невнимательности

i2GR committed:
SP9 FP review 2 fix
fix based review 2 comments

by Igor I.

afomkina reviewed (approved):
Отлично справился 🔥
